pkg/controller/route/routecontroller.go

```diff
@@ -33,79 +43,217 @@ type RouteController struct {
 	kubeClient  client.Interface
 	clusterName string
 	clusterCIDR *net.IPNet
+
+	// To allow injection of syncRoutine for testing.
+	syncHandler func(key string) error
+
+	// Node framework and store
+	nodeController *framework.Controller
+	nodeStore      cache.StoreToNodeLister
+
+	// routes that need to be created
+	queue *workqueue.Type
 }
 
-func New(routes cloudprovider.Routes, kubeClient client.Interface, clusterName string, clusterCIDR *net.IPNet) *RouteController {
-	return &RouteController{
+func NewRouteController(routes cloudprovider.Routes, client client.Interface, resyncPeriod controller.ResyncPeriodFunc, clusterName string, clusterCIDR *net.IPNet) *RouteController {
+	rc := &RouteController{
 		routes:      routes,
-		kubeClient:  kubeClient,
+		kubeClient:  client,
 		clusterName: clusterName,
 		clusterCIDR: clusterCIDR,
+		queue:       workqueue.New(),
 	}
+
+	rc.nodeStore.Store, rc.nodeController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options api.ListOptions) (runtime.Object, error) {
+				return rc.kubeClient.Nodes().List(options)
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return rc.kubeClient.Nodes().Watch(options)
+			},
+		},
+		&api.Node{},
+		resyncPeriod(),
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: rc.enqueueNode,
+			UpdateFunc: func(old, cur interface{}) {
+				oldNode := old.(*api.Node)
+				curNode := cur.(*api.Node)
+				// if PodCIDR of node changed, we need recreate the route
+				if oldNode.Spec.PodCIDR != curNode.Spec.PodCIDR {
+					glog.V(4).Infof("Observed PodCIDR updated for node: %v, %v->%v", curNode.Name, oldNode.Spec.PodCIDR, curNode.Spec.PodCIDR)
+					rc.enqueueNode(cur)
+				}
+			},
+			DeleteFunc: rc.enqueueNode,
```

Comment:
Having Add, Update, and Delete all use the same function makes the function necessarily more complicated. Do we need to do it this way? Is there a way to organize it so that we have three easy to understand helper functions?